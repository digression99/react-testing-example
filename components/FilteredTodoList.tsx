import {Box, Text, List, ListItem, Button, Input, Menu, MenuButton, MenuList, MenuItem} from '@chakra-ui/react'
import { faker } from '@faker-js/faker'
import {useCallback, useState, ComponentType, useMemo} from "react";
import { ChevronDownIcon } from '@chakra-ui/icons'

type Todo = {
  title: string
  completed: boolean
  id: string
}
type TodoFilterType = 'ALL' | 'COMPLETED' | 'WAITING'

const getTodoId = () => faker.datatype.uuid()
const getToggledTodo = (id: string, todo: Todo) => todo.id === id ? { ...todo, completed: !todo.completed } : todo
const getFilteredTodo = (filterType: TodoFilterType, todo: Todo) =>
  filterType === 'COMPLETED' && todo.completed
  || filterType === 'WAITING' && !todo.completed
  || filterType === 'ALL'

const useTodos = () => {
  const [list, setList] = useState<Todo[]>([])
  const [filter, setFilter] = useState<TodoFilterType>('ALL')

  const addTodo = useCallback(
    (title: string) => setList(ls =>
      ([...ls, {
        title, completed: false, id: getTodoId()
      }])
    ),
    [])

  const removeTodo = useCallback((id: string) =>
    setList(ls =>
      ls.filter(todo =>
        todo.id !== id
      )), [])

  const toggleTodo = useCallback((id: string) =>
    setList(ls =>
      ls.map(todo => getToggledTodo(id, todo)
      )), [])

  const changeFilter = useCallback((filterType: TodoFilterType) => setFilter(filterType), [])
  const filteredList = useMemo(() => list.filter(todo => getFilteredTodo(filter, todo)), [filter, list])
  return { list: filteredList, addTodo, removeTodo, toggleTodo, changeFilter }
}

type PureProps = {
  list: Todo[]
  onToggleTodo: (id: string) => void
  onRemoveTodo: (id: string) => void
  onAddTodo: (title: string) => void
  onFilterTodo: (filterType: TodoFilterType) => void
}

const PureFilteredTodoList = ({ list, onToggleTodo, onRemoveTodo, onAddTodo, onFilterTodo }: PureProps) => {
  const [title, setTitle] = useState('')

  const onCreate = () => {
    onAddTodo(title)
    setTitle('')
  }

  return (
    <Box>
      <Box>
        <Text>Todo Form</Text>
        <Input value={title} onChange={e => setTitle(e.target.value)} />
        <Button onClick={onCreate} disabled={!title}>Create</Button>
      </Box>

        <Menu>
          <MenuButton as={Button} rightIcon={<ChevronDownIcon />}>
            Filters
          </MenuButton>
          <MenuList>
            <MenuItem onClick={() => onFilterTodo('ALL')}>All</MenuItem>
            <MenuItem onClick={() => onFilterTodo('COMPLETED')}>Completed</MenuItem>
            <MenuItem onClick={() => onFilterTodo('WAITING')}>Waiting</MenuItem>
          </MenuList>
        </Menu>

      <List>
        {list.map(todo =>
          <ListItem key={todo.id}>
            <Text textDecoration={todo.completed ? 'line-through' : 'none'}>{todo.title}</Text>
            <Box>
              <Button onClick={() => onRemoveTodo(todo.id)}>Remove</Button>
              <Button onClick={() => onToggleTodo(todo.id)}>Toggle</Button>
            </Box>
          </ListItem>)}
      </List>
    </Box>
  )
}

type Props = {}

const withTodos = (Component: ComponentType<PureProps>) => {
  return (props: Props) => {
    const { list, addTodo, removeTodo, toggleTodo, changeFilter } = useTodos()
    return <Component
      {...props}
      list={list}
      onAddTodo={addTodo}
      onFilterTodo={changeFilter}
      onRemoveTodo={removeTodo}
      onToggleTodo={toggleTodo}
    />
  }
}

export const FilteredTodoList = withTodos(PureFilteredTodoList)
