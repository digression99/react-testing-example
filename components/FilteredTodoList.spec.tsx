import { render, screen } from '@testing-library/react'
import { FilteredTodoList, PureFilteredTodoList, Todo } from './FilteredTodoList'
import userEvent from '@testing-library/user-event'

const fakeData: Todo[] = [
  { title: 'Test todo 1', completed: false, id: 'test-id-1' },
  { title: 'Test todo 2', completed: true, id: 'test-id-2' },
  { title: 'Test todo 3', completed: false, id: 'test-id-3' },
  { title: 'Test todo 4', completed: true, id: 'test-id-4' },
]

describe('<FilteredTodoList />', () => {
  test('renders the component correctly', () => {
    const { container } = render(
      <PureFilteredTodoList
        list={fakeData}
        onToggleTodo={jest.fn()}
        onRemoveTodo={jest.fn()}
        onFilterTodo={jest.fn()}
        onAddTodo={jest.fn()}
      />
    )
    expect(container).toMatchSnapshot()
  })

  test('Add todo with a text, and it calls the method.', async () => {
    const user = userEvent.setup()
    const addTodoFn = jest.fn()
    render(
      <PureFilteredTodoList
        list={fakeData}
        onToggleTodo={jest.fn()}
        onRemoveTodo={jest.fn()}
        onFilterTodo={jest.fn()}
        onAddTodo={addTodoFn}
      />
    )
    const input = screen.getByRole('textbox', { name: 'title' })
    await user.type(input, 'Test todo 5')
    const createButton = screen.getByRole('button', { name: 'Create' })
    await user.click(createButton)
    expect(addTodoFn).toBeCalledWith('Test todo 5')
  })

  test('Add todo with a text, and it shows on the list', async () => {
    const user = userEvent.setup()
    render(<FilteredTodoList />)

    const input = screen.getByRole('textbox', { name: 'title' })
    await user.type(input, 'Test todo 1')
    const createButton = screen.getByRole('button', { name: 'Create' })
    await user.click(createButton)
    expect(screen.getByText('Test todo 1')).toBeInTheDocument()
  })

  test('Change filter to "Completed" and it changes the list', async () => {
    const user = userEvent.setup()
    render(<FilteredTodoList />)

    const input = screen.getByRole('textbox', { name: 'title' })
    await user.type(input, 'Test todo 1')
    const createButton = screen.getByRole('button', { name: 'Create' })
    await user.click(createButton)
    expect(screen.getByText('Test todo 1')).toBeInTheDocument()

    await user.type(input, 'Test todo 2')
    await user.click(createButton)
    expect(screen.getByText('Test todo 2')).toBeInTheDocument()

    const menu = await screen.findByText(/filters/i)
    await user.click(menu)
    const completedFilterButton = await screen.findByText(/completed/i)
    await user.click(completedFilterButton)
    expect(screen.queryAllByText(/Test todo/i)).toHaveLength(0)

    await user.click(menu)
    const allFilterButton = await screen.findByText(/all/i)
    await user.click(allFilterButton)

    // You can debug the screen with this method.
    screen.debug()
    expect(screen.getByText('Test todo 1')).toBeInTheDocument()
    expect(screen.queryAllByText(/Test todo/i)).toHaveLength(2)
  })
})
