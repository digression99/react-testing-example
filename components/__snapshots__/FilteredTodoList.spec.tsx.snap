// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilteredTodoList /> renders the component correctly 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-0"
    >
      <p
        class="chakra-text css-0"
      >
        Todo Form
      </p>
      <input
        aria-label="title"
        class="chakra-input css-0"
        value=""
      />
      <button
        class="chakra-button css-0"
        disabled=""
        type="button"
      >
        Create
      </button>
    </div>
    <button
      aria-controls="menu-list-:r1:"
      aria-expanded="false"
      aria-haspopup="menu"
      class="chakra-button chakra-menu__menu-button css-0"
      id="menu-button-:r1:"
      type="button"
    >
      <span
        class="css-0"
      >
        Filters
      </span>
      <span
        class="chakra-button__icon css-0"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-0"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            fill="currentColor"
          />
        </svg>
      </span>
    </button>
    <div
      class="css-0"
      style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
    >
      <div
        aria-orientation="vertical"
        class="chakra-menu__menu-list css-0"
        id="menu-list-:r1:"
        role="menu"
        style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
        tabindex="-1"
      >
        <button
          class="chakra-menu__menuitem css-0"
          data-index="0"
          id="menu-list-:r1:-menuitem-:r2:"
          role="menuitem"
          tabindex="-1"
          type="button"
        >
          All
        </button>
        <button
          class="chakra-menu__menuitem css-0"
          data-index="1"
          id="menu-list-:r1:-menuitem-:r3:"
          role="menuitem"
          tabindex="-1"
          type="button"
        >
          Completed
        </button>
        <button
          class="chakra-menu__menuitem css-0"
          data-index="2"
          id="menu-list-:r1:-menuitem-:r4:"
          role="menuitem"
          tabindex="-1"
          type="button"
        >
          Waiting
        </button>
      </div>
    </div>
    <ul
      class="css-0"
      role="list"
    >
      <li
        class="css-0"
      >
        <p
          class="chakra-text css-0"
        >
          Test todo 1
        </p>
        <div
          class="css-0"
        >
          <button
            class="chakra-button css-0"
            type="button"
          >
            Remove
          </button>
          <button
            class="chakra-button css-0"
            type="button"
          >
            Toggle
          </button>
        </div>
      </li>
      <li
        class="css-0"
      >
        <p
          class="chakra-text css-0"
        >
          Test todo 2
        </p>
        <div
          class="css-0"
        >
          <button
            class="chakra-button css-0"
            type="button"
          >
            Remove
          </button>
          <button
            class="chakra-button css-0"
            type="button"
          >
            Toggle
          </button>
        </div>
      </li>
      <li
        class="css-0"
      >
        <p
          class="chakra-text css-0"
        >
          Test todo 3
        </p>
        <div
          class="css-0"
        >
          <button
            class="chakra-button css-0"
            type="button"
          >
            Remove
          </button>
          <button
            class="chakra-button css-0"
            type="button"
          >
            Toggle
          </button>
        </div>
      </li>
      <li
        class="css-0"
      >
        <p
          class="chakra-text css-0"
        >
          Test todo 4
        </p>
        <div
          class="css-0"
        >
          <button
            class="chakra-button css-0"
            type="button"
          >
            Remove
          </button>
          <button
            class="chakra-button css-0"
            type="button"
          >
            Toggle
          </button>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
